lib/spack/spack/version/version_types.py

```diff
@@ -61,27 +62,27 @@ def __str__(self):
     def __repr__(self) -> str:
         return f'VersionStrComponent("{self}")'
 
-    def __eq__(self, other):
+    def __eq__(self, other) -> bool:
```

Comment:
Maybe:
```suggestion
    def __eq__(self, other: object) -> bool:
```
? Also, I wonder if a slightly better implementation of this is:
```
def __eq__(self, other) -> bool:
    if not isinstance(other, VersionStrComponent):
        return NotImplemented
    return self.data == other.data
```
i.e. do not compare against other types.

Reply:
I am unclear on whether it's supposed to be a best practice to annotate arguments that are just supposed to be `object` -- is it?  I think adding `-> bool` is sufficient to get the typing checked, and annotating as `object` just adds noise to the dunder methods inherited from `object`.  I couldn't find anything saying what we should be doing this.

For comparing other objects I would lean towards not causing too many failures - as long as `__eq__` is `False` for other object types I think it's fine.... I don't know that limiting things helps a lot.  What do you think?

Reply:
I dug into annotating this and the only clean-ish version I could come up with that mypy supports is this:

```python
    def __lt__(self, other: object) -> bool:
        lhs_inf = isinstance(self.data, int)
        if isinstance(other, int):
            return not lhs_inf

        if not isinstance(other, VersionStrComponent):
            raise TypeError(f"'<' not supported for instances of {type(other)}")

        rhs_inf = isinstance(other.data, int)
        lhs_t = (self.data, "") if lhs_inf else (-1, self.data)
        rhs_t = (other.data, "") if rhs_inf else (-1, other.data)
        return lhs_t < rhs_t
```

Which I find much less elegant and probably less efficient than what we had before, and I'm not inclined to change the function's performance as part of this PR. Instead, I put a note and some ignore lines on these. I think we can revisit if mypy gets smarter somehow, and if we can show that the extra type check wouldn't slow these comparisons down.

Reply:
That's fine with me. For reference, I found this discussion: https://discuss.python.org/t/make-type-hints-for-eq-of-primitives-less-strict/34240 Type-hinting dunder methods seems a bit controversial at the moment.